ARE/middleware/src/main/java/eu/asterics/mw/are/BundleManager.java

```diff
@@ -132,16 +144,173 @@ public BundleManager(ModelValidator modelValidator) {
+                logger.fine("bundle change: false");
+                return false;
+            }
+        }
+        
+        logger.fine("bundle change: true");
+        return true;
+    }
+
+    /**
+     * Stores the given MD5 hash hex string to a file.
+     * @param md5Sum
+     * @throws URISyntaxException
+     * @throws IOException
+     */
+    private void storeMD5Sum(String md5Sum) throws URISyntaxException, IOException {
+        logger.fine("Storing MD5 sum...");
+        URI pluginJarsHashFileURI=ResourceRegistry.getInstance().getResource(CACHING_PLUGIN_JARS_HASH_TXT, RES_TYPE.TMP);
+        File pluginJarsHashFile=ResourceRegistry.toFile(pluginJarsHashFileURI);
+        File parentFile=pluginJarsHashFile.getParentFile();
+        
+        if(!parentFile.exists()) {
+            if(!parentFile.mkdirs()) {
+                throw new IOException("Could not create parent directories for MD5 hash file: "+pluginJarsHashFile);
+            }
+        }
+        try(FileWriter writer=new FileWriter(pluginJarsHashFile)) {
+            writer.write(md5Sum.trim());
+        }
+    }
+
+    /**
+     * This method generates all needed cachde files.
+     * @throws MalformedURLException
+     * @throws IOException
+     * @throws ParseException
+     * @throws URISyntaxException
+     */
+    private void generateCacheFiles() throws MalformedURLException, IOException, ParseException, URISyntaxException {        
+        logger.fine("Generating cache files/help files");
+        long startTime=System.currentTimeMillis();       
+        //Before we can start, let's ensure the generation of cached files
+        generateLoaderComponentListAndBundleDescriptorCacheFiles();
+        logger.fine(MessageFormat.format("Generated cache files in {0} ms",(System.currentTimeMillis()-startTime)));
+    }
+
+    /**
+     * Generates an MD5 hex string of the bytes of all plugin jars.
+     * A plugin jar is defined as an osgi bundle containing a bundle_descriptor.xml file in the root folder of the bundle.
+     * @return
+     * @throws NoSuchAlgorithmException
+     * @throws MalformedURLException
+     * @throws IOException
+     */
+    private String generateMD5Sum() throws NoSuchAlgorithmException, MalformedURLException, IOException {
+        logger.fine("Generating MD5 sum of bundle jars....");
+        long timeStart=System.currentTimeMillis();
+        MessageDigest MDInstance=MessageDigest.getInstance("MD5");
+        List<URI> compJarList=ResourceRegistry.getInstance().getComponentJarList(false);
+        
+        for(URI compJarURI : compJarList) {
+            try(DigestInputStream dis=new DigestInputStream(compJarURI.toURL().openStream(), MDInstance)) {
+                //Just opening automatically updates the MessageDigest object.
+                byte[] readBytes = new byte[1024];
+                int numRead=0;
+                while((numRead=dis.read(readBytes))!=-1) {
+                    MDInstance.update(readBytes,0,numRead);
```

Comment:
I do not think you need to update the `MDInstance` here. You are using a `DigestInputStream` that does this automatically. Actually I think this way you are calculating a wrong MD5 hash because all bytes are added two times to the MessageDigest.

Reply:
instead of calculating the hash "manually" here you could also use:
https://commons.apache.org/proper/commons-codec/apidocs/org/apache/commons/codec/digest/DigestUtils.html#digestAsHex-java.io.InputStream-
I think you added the commons-codec package with this merge?!

Reply:
regarding calculation: you are right, it also works without calling update

I did not want to use digestAsHex here because I wanted to add all bytes of all plugin jars to one instance of DigestInputStream and there is no method for such a case.